### PR TITLE
feat: adding script to automate folder structure

### PR DESCRIPTION
**Pull request by himanshu010**
fixes: #1310 

**What kind of change does this PR introduce?**

feature
**Did you add tests for your changes?**
N/A
**If relevant, did you update the documentation?**
yes
**Summary**
It will automate the folder-structure.md file so that it will change on renaming the files.

**Does this PR introduce a breaking change?**

**Other information**

### CODE REVIEW

**Comment by himanshu010** [comment]
Reviewers can check the working of this [here](https://github.com/himanshu010/folder-structure)..

File: folder-structure.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
+var fs = require('fs');
+var path = require('path');
+var util = require('util');
+
+var logFile = fs.createWriteStream('folder-structure.md', { flags: 'w' });
+var logStdout = process.stdout;
+
+console.log = function() {
```
Comment (ematipico) [inline folder-structure.js]: Please don't override built in methods. You can create a new function

File: folder-structure.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
+var fs = require('fs');
```
Comment (ematipico) [inline folder-structure.js]: Change `var` to `const`

File: folder-structure.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+var logFile = fs.createWriteStream('folder-structure.md', { flags: 'w' });
+var logStdout = process.stdout;
+
+console.log = function() {
+    logFile.write(util.format.apply(null, arguments) + '\n');
+    logStdout.write(util.format.apply(null, arguments) + '\n');
+};
+console.error = console.log;
+
+var FolderNamesArray = [];
```
Comment (ematipico) [inline folder-structure.js]:
```suggestion
var folderNamesArray = [];
```

File: folder-structure.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+function crawl(spacing, dir, level) {
+    var folderName = path.basename(dir);
+    FolderNamesArray.push(folderName);
+
+    console.log(spacing + "├───" + folderName);
+
+    var files = fs.readdirSync(dir);
+    for (var x in files) {
+        //ignored folders
+        if (files[x] !== '.git' && files[x] !== '.nyc_output'&& files[x] !== 'coverage'&& files[x] !== 'node_modules'&& files[x] !== 'bin'&& files[x] !== 'lib') {
```
Comment (ematipico) [inline folder-structure.js]: Put the folder names you don't want to check inside an object:

```js
const blacklist = {
  '.git': true,
  'lib': true,
  // ... etc
}
```
And here you can do

```js
blacklist[files[x]]
```

**Comment by ematipico** [comment]
I think one piece missing here, it's the part of WHEN this script is called. Do you have any idea of when a new folder structure should be generated/updated?

**Comment by himanshu010** [comment]
> I think one piece missing here, it's the part of WHEN this script is called. Do you have any idea of when a new folder structure should be generated/updated?

I think that the renaming of directories is not very often. So we can mention this in the contributing.md file like this:
Screenshot (63)

**Comment by ematipico** [comment]
I strongly disagree with this. In a world where everything can be automated, you're asking contributors and maintainers to run this script every time they do a change. We are basically back to the same issue where someone has to remember to update a readme file when changing the file structure.

**Comment by snitin315** [comment]
Renaming is not often but new directories are continuously added in the project , with each PR the folder structures changes, So we don't want to run the script everytime we create a new file.

**Comment by himanshu010** [comment]
> I strongly disagree with this. In a world where everything can be automated, you're asking contributors and maintainers to run this script every time they do a change. We are basically back to the same issue where someone has to remember to update a readme file when changing the file structure.

okay, so then I think that running this script on each commit will do that task. What are your views on this @ematipico

**Comment by ematipico** [comment]
I am still convinced that we don't really need this. But other people think it might be useful, so I suggest other people to come up with a solution to avoid less disruption as possible

**Comment by maria99williams** [comment]
I think that we are just making folder structure more and more complex with every step. Folder structure is just a .md file like the readme file which just needs to be updated when you do a major change in the repo like the readme.

**Comment by ematipico** [comment]
@maria99williams but would you need only an overview of the folder structure (e.g. "src" contains source code of the module, "lib" the compiled code, etc.) or a detailed one where we list all the folders?

**Comment by snitin315** [comment]
Right now we are in beta version , the files and folder structure is continuously changing also theese type of folder structure makes more sense for projects like `create-react-app` , where   the user has to know about it . 
I'm not sure if it is required for CLI tools , makes more sense with Front-end apps kind of projects.

**Comment by himanshu010** [comment]
> Right now we are in beta version , the files and folder structure is continuously changing

I think that any action with the files does not affect the folder structure as it is only affected by the folders and yeah directories are created not very often. So i think that we can try this in the beta version repo and if it affects the repo more which i think have a very less probablity we can just remove it

**Comment by maria99williams** [comment]
@ematipico detailed structure will be good as it tells the description as well as the location and location of the file can tell many things on its own

**Comment by snitin315** [comment]
You can check some of last merged PR's , for each feature there is atleast an increased test directory. Even some standalone packages are planned to be added , that will create a great impact on folder structure.

**Comment by snitin315** [comment]
> and location of the file can tell many things on its own

It will be great if you can list down some of the many things.

**Comment by maria99williams** [comment]
Like different packages can have files with same name which may or may not have the same functions, in this case location can help

**Comment by snitin315** [comment]
As far as I have gone through the source code we have used meaningful and easy to understand names for files. you can guess the functionality of file with its name like `cli-flags` will contain all flags passed to CLI, `utils` will contain utilities, OutputGroup will contain arguments related to `output` etc. 

Rest are universal names like `src`, `lib`, `test`, `__tests__` .

**Comment by snitin315** [comment]
My point is as a cli user I don't care about the source code I can download and use it.

As a contributor the folder structure is very much self explainatory

**Comment by himanshu010** [comment]
> As a contributor the folder structure is very much self explainatory

I submitted a issue in which I discussed the idea and many contributors said that having a folder structure

**Comment by snitin315** [comment]
> I submitted an issue in which I discussed the idea and many contributors said that having a folder structure will help and therefore I started to work on this

I know even it was convincing to me as well at first but the following points changed my mind.

- Version not stable (breaking changes with continuously changing Folder Structure). features are being added and removed as well. 

-  [comment](https://github.com/webpack/webpack-cli/pull/1312#issuecomment-597633046)

-   Makes more sense for projects like `create-react-app`

**Comment by himanshu010** [comment]
I am not saying that you have to merge the pr @ematipico you can close it if you want. I am just saying that I worked as the issue progressed and fulfilled all the requirements as the issue furt

**Comment by himanshu010** [comment]
> * Version not stable (breaking changes with continuously changing Folder Structure). features are being added and removed as well.

Now you don't have to refresh the folder-structure file, it will refresh automatically according to the latest naming. See #1320

**Comment by snitin315** [comment]
> Now you don't have to refresh the folder-structure file, it will refresh automatically according to the latest naming. See #1320

valid for `yarn commit` or `npm run commit` only .

Also, this change doesn't require a separate PR. please try to avoid confusion. you have created 3 different issues & 3diffreent PR's for solving a single issue. This is a very very bad practice. Please avoid.

**Comment by himanshu010** [comment]
> valid for yarn commit or npm run commit only

This is the authorised way to make a pr. And a contributor must have to follow the steps written in contributing.md. So this will definitely work if the commit is valid

**Comment by snitin315** [comment]
>This is the authorised way to make a pr. 
Not necessarily until you are writing commits as required.

>And a contributor must have to follow the steps written in contributing.md. 

It's you who is not following the Contributing.md ,
( P.S. check the branching policy. )

**Comment by himanshu010** [comment]
> It's you who is not following the Contributing.md 

I think bro the GSoC fever has taken charge on you and you are just seeking more and more topics to argue and to nullify the work of other contributors.  The Reviewing team is there for a reason, let them do their work.
And please don't be rude to a new contributor here. Your previous comment is very rude for a respected community like this. We all are here to make this project better. If you don't like a pr you can simply vote for closing it.

**Comment by snitin315** [comment]
> And please don't be rude to a new contributor here

I'm not being rude sorry if you felt like that. I am just expressing my points (That's what we do in open source.)

**Comment by himanshu010** [comment]
merged in #1321

**Pull request closed without merging**